cmd/basic/main.go

```diff
@@ -78,152 +82,204 @@ func main() {
+
+			valSeed := new(big.Int).SetBytes([]byte(key)).Int64()
+			subRS.Seed(valSeed)
+			value := mct.RandBytes(&subRS, int(conf.ValueSize))
+
+			_, err := mc.Set(key, uint32(conf.ClientFlags), uint32(conf.KeyTTL), value)
+			if err != nil {
+				fmt.Println(err)
+				return err
+			}
+		}
 	}
+
+	return nil
 }
 
-// TODO: use sync.Pool for Item/etc?
-// pool.Put() items back before sleep.
-// may also be able to cache mc's bufio's this way.
-func (l *BasicLoader) Worker(doneChan chan<- int) {
-	// FIXME: selector.
-	host := l.servers[0]
-	mc := mct.NewClient(host)
-	bundles := l.requestBundlesPerConn
+func (conf *Config) Worker(index int, results chan Stats) error {
+	workerSeed := conf.RngSeed + int64(index) + int64(conf.KeySpace)
+	stats := Stats{}
+	mc := mct.NewClient(conf.Servers[0], conf.Socket, conf.Pipelines, conf.KeyPrefix, conf.StripKeyPrefix)
 
-	rs := pcgr.New(time.Now().UnixNano(), 0)
+	rs := pcgr.New(workerSeed, 0)
+	randR := rand.New(&rs)
 	var zipRS *rand.Zipf
-	randR := rand.New(&rs) // main randomizer, so we can use the random interface.
-	if l.useZipf {
-		zipRS = rand.NewZipf(randR, l.zipfS, l.zipfV, uint64(l.keySpace))
+	if conf.UseZipf {
+		zipRS = rand.NewZipf(randR, conf.ZipfS, conf.ZipfV, uint64(conf.KeySpace))
 		if zipRS == nil {
-			fmt.Printf("bad arguments to zipf: S: %f V: %f\n", l.zipfS, l.zipfV)
-			return
+			fmt.Printf("bad arguments to zipf: S: %f V: %f\n", conf.ZipfS, conf.ZipfV)
+			return nil
 		}
 	}
+	subRS := pcgr.New(1, 0)
+
+	var rl ratelimit.Limiter
+	if conf.RPS > 0 {
+		rl = ratelimit.New(conf.RPS)
+	} else {
+		rl = ratelimit.NewUnlimited()
+	}
 
-	subRS := pcgr.New(1, 0) // randomizer is re-seeded for random strings.
-	var res int
-	defer func() { doneChan <- res }()
+	for start := time.Now(); ; {
+		iterStart := time.Now()
+		if iterStart.Sub(start) > conf.Duration {
+			break
+		}
 
-	for bundles == -1 || bundles > 0 {
-		if bundles != -1 {
-			bundles--
+		if conf.UseZipf {
+			subRS.Seed(int64(zipRS.Uint64()))
+		} else {
+			subRS.Seed(conf.RngSeed + int64(randR.Intn(conf.KeySpace)))
 		}
-		for i := l.requestsPerSleep; i > 0; i-- {
-			// generate keys
-			// TODO: Allow min/max length for keys.
-			// The random key needs to stick with the random length, or we end
-			// up with keySpace * (max-min) number of unique keys.
-			// Need to pull the randomizer exactly once (then just modulus for
-			// a poor-but-probably-fine random value), then build the random
-			// string from the rest.
-			// Could also re-seed it twice, pull once Intn for length,
-			// re-seed, then again for key space.
-
-			keyLen := l.keyLength
-			if l.useZipf {
-				subRS.Seed(int64(zipRS.Uint64()))
-			} else {
-				subRS.Seed(int64(randR.Intn(l.keySpace)))
+
+		key := mct.RandString(&subRS, conf.KeyLength, conf.KeyPrefix)
+		valSeed := new(big.Int).SetBytes([]byte(key)).Int64()
+		subRS.Seed(valSeed)
+
+		switch rng := randR.Intn(conf.DelRatio + conf.SetRatio + conf.GetRatio); {
```

Comment:
Core worker logic simply branches probabilistically according to the delete-set-get ratios (respectively).